lib/charms/synapse/v0/matrix_auth.py

```diff
@@ -0,0 +1,516 @@
+class MatrixAuthProviderData(BaseModel):
+    """Represent the MatrixAuth provider data.
+
+    Attributes:
+        homeserver: the homeserver URL.
+        shared_secret: the Matrix shared secret.
+        shared_secret_id: the shared secret Juju secret ID.
+    """
+
+    homeserver: Optional[str] = None
+    shared_secret: Optional[SecretStr] = Field(default=None, exclude=True)
+    shared_secret_id: Optional[SecretStr] = Field(default=None)
+
+    def set_shared_secret_id(self, model: ops.Model, relation: ops.Relation) -> None:
+        """Store the Matrix shared secret as a Juju secret.
+
+        Args:
+            model: the Juju model
+            relation: relation to grant access to the secrets to.
+        """
+        # password is always defined since pydantic guarantees it
+        password = cast(SecretStr, self.shared_secret)
+        # pylint doesn't like get_secret_value
+        secret_value = password.get_secret_value()  # pylint: disable=no-member
+        try:
+            secret = model.get_secret(label="shared-secret")
+            secret.set_content({"shared-secret-content": secret_value})
+            # secret.id is not None at this point
+            self.shared_secret_id = cast(str, secret.id)
+        except ops.SecretNotFoundError:
+            secret = relation.app.add_secret(
+                {"shared-secret-content": secret_value}, label="shared-secret"
+            )
+            secret.grant(relation)
+            self.shared_secret_id = cast(str, secret.id)
+
+    @classmethod
+    def get_shared_secret(
+        cls, model: ops.Model, shared_secret_id: Optional[str]
+    ) -> Optional[SecretStr]:
+        """Retrieve the shared secret corresponding to the shared_secret_id.
+
+        Args:
+            model: the Juju model.
+            shared_secret_id: the secret ID for the shared secret.
+
+        Returns:
+            the shared secret or None if not found.
+        """
+        if not shared_secret_id:
+            return None
+        try:
+            secret = model.get_secret(id=shared_secret_id)
+            password = secret.get_content().get("shared-secret-content")
+            if not password:
+                return None
+            return SecretStr(password)
+        except ops.SecretNotFoundError:
+            return None
+
+    def to_relation_data(self, model: ops.Model, relation: ops.Relation) -> Dict[str, str]:
+        """Convert an instance of MatrixAuthProviderData to the relation representation.
+
+        Args:
+            model: the Juju model.
+            relation: relation to grant access to the secrets to.
+
+        Returns:
+            Dict containing the representation.
+        """
+        self.set_shared_secret_id(model, relation)
+        dumped_model = self.model_dump(exclude_unset=True)
+        dumped_data = {
+            "shared_secret_id": dumped_model["shared_secret_id"],
+            "homeserver": dumped_model["homeserver"],
```

Comment:
What will be the format?
Maubot, for example, expects something like http://10.1.74.65:8080

Reply:
Using server_name might be challenging in case of different URL (ubuntu.com x chat-server.ubuntu.com)

What about adding a config to set the URL that will be used for the integration? Default to server_name

Reply:
That's a fair point, the server name is indeed ubuntu.com and while Element clients look for a .wellknown file under the domain, plugins don't do that (looking at the irc plugin code it will not go that extra length of searching for a .wellknown file, it will connect to the name you give it). That's a great catch! I think the safe bet here is to have it as a config. I don't think we can get the DNS entry for the Synapse homeserver from Synapse, we can extract the homeserver, though clients resolve it again...

Reply:
+1 for config then :)